components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java

```diff
@@ -1,42 +1,115 @@
+        try {
+            APIConsumer apiConsumer = RestApiUtil.getConsumer(username);
+            if (!StringUtils.isEmpty(apiId)) {
+                APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifier(apiId);
+                subscriptions = apiConsumer.getSubscribedIdentifiers(subscriber, apiIdentifier, groupId);
+
+            } else if (!StringUtils.isEmpty(applicationId)) {
+                Application application = apiConsumer.getApplicationById(applicationId);
+                subscriptions =
+                        apiConsumer.getSubscribedAPIs(subscriber, application.getName(), application.getGroupId());
+            }
+
+            List<SubscriptionDTO> subscriptionDTOs = new ArrayList<>();
+            for (SubscribedAPI subscription : subscriptions) {
+                SubscriptionDTO subscriptionDTO = SubscriptionMappingUtil.fromSubscriptiontoDTO(subscription);
+                //when retrieving subscriptions from getSubscribedAPIs(), existing application ids are not populated
+                if (!StringUtils.isEmpty(applicationId)) {
+                    subscriptionDTO.setApplicationId(applicationId);
+                }
+                subscriptionDTOs.add(subscriptionDTO);
+            }
+            return Response.ok().entity(subscriptionDTOs).build();
+
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
+        }
     }
+
     @Override
-    public Response subscriptionsSubscriptionIdGet(String accept,String ifNoneMatch,String ifModifiedSince){
-        // do some magic!
-        return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, "magic!")).build();
+    public Response subscriptionsPost(SubscriptionDTO body, String contentType) {
+        String username = RestApiUtil.getLoggedInUsername();
+        APIConsumer apiConsumer = null;
+        try {
+            //todo: Validation for allowed throttling tiers and Tenant based validation for subscription
+            apiConsumer = RestApiUtil.getConsumer(username);
+            String apiId = body.getApiId();
+            String applicationId = body.getApplicationId();
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifier(apiId);
+            apiIdentifier.setTier(body.getTier());
+            int subscriptionId =
+                    apiConsumer.addSubscription(apiIdentifier, username, Integer.parseInt(applicationId));
+            SubscribedAPI addedSubscribedAPI = apiConsumer.getSubscriptionById(subscriptionId);
+            SubscriptionDTO addedSubscriptionDTO = SubscriptionMappingUtil.fromSubscriptiontoDTO(addedSubscribedAPI);
+            //todo: use a proper way other than using "subscriptions/"
+            return Response.created(new URI("subscriptions/" + subscriptionId)).entity(addedSubscriptionDTO).build();
+        } catch (APIManagementException | URISyntaxException e) {
+            throw new InternalServerErrorException(e);
```

Comment:
Exception handling already fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7) and #1533 (https://github.com/wso2/carbon-apimgt/commit/243966339e63e60a915cb07ee4e5562a9cb7facf)